Add Post Freezing Optimizations, turn on by default in torch.jit.freeze

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50241 Re-enable no-op peepholes
* **#50222 Add Post Freezing Optimizations, turn on by default in torch.jit.freeze**
* #50221 Peephole Optimize out conv(x).dim(), which prevents BN fusion
* #50220 [JIT] Factor out peephole to own test file
* #50075 [JIT] Add Frozen Conv-> Add/Sub/Mul/Div fusion
* #50074 [JIT] Frozen Graph Conv-BN fusion

This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in \~prototype\~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations.

Differential Revision: [D25856264](https://our.internmc.facebook.com/intern/diff/D25856264)

Commit:
Add Post Freezing Optimizations, turn on by default in torch.jit.freeze

[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"




[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in ~prototype~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations. 









[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in ~prototype~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations. 









[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in ~prototype~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations. 









[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in ~prototype~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations. 









[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in ~prototype~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations. 









[ghstack-poisoned]

bzinodev reviewed (commented):
How does it interact with graph mode quantization? And also for mobile? I suggest to have couple a test?

eellison commented:
It doesn't, because they all use  _freeze_module, not the torch.jit.freeze API. If they did  tests would break.

ZolotukhinM commented on torch/csrc/jit/passes/frozen_graph_optimizations.cpp:
```diff
@@ -0,0 +1,21 @@
+namespace jit {
+
+void OptimizeFrozenGraph(std::shared_ptr<Graph>& graph) {
+  // run a couple times to capture Conv -> Mul -> Add etc
```
Might be worth running that while something is changed (with some threshold to not run for too long).

eellison replied:
This would be nice but i'm not sure how relevant it is, might do as a follow up.

ZolotukhinM replied:
I mentioned this simply because it's a common pattern in compilers. I agree with you that it's not clear whether it matters here.

ZolotukhinM commented on torch/jit/_freeze.py:
```diff
@@ -10,7 +10,7 @@
 
 
-def freeze(mod, preserved_attrs: Optional[List[str]] = None):
+def freeze(mod, preserved_attrs: Optional[List[str]] = None, optimize: bool = True):
```
It might be easier to land this in two steps: 1) add a new flag with the new functionality, but disabled by default, 2) flip the default. This way if something breaks, only the flag switch would be reverted.

ZolotukhinM reviewed (approved):
Looks good to me!

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in \~prototype\~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations. 









[ghstack-poisoned]

Commit:
Update on "Add Post Freezing Optimizations, turn on by default in torch.jit.freeze"


This PR adds a pass which runs a set of optimizations to be done after freezing. Currently this encompasses Conv-BN folding, Conv->Add/Sub/Mul/Div folding and i'm also planning on adding dropout removal. 

I would like some feedback on the API. torch.jit.freeze is technically in \~prototype\~ phase so we have some leeway around making changes. I think in the majority of cases, the user is going to want to freeze their model, and then run in inference. I would prefer if the optimization was opt-out instead of opt-in. All internal/framework use cases of freezing all use `freeze_module`, not the python API, so this shouldn't break anything. 

I have separated out the optimization pass as a separate API to make things potentially modular, even though I suspect that is an unlikely case. In a future PR i would like to add a `torch::jit::freeze` which follows the same api as `torch.jit.freeze` intended for C++ use, and runs the optimizations.

Differential Revision: [D25856264](https://our.internmc.facebook.com/intern/diff/D25856264)

[ghstack-poisoned]

xsacha commented:
@eellison as a suggestion for another post freezing optimisation: you could convert the style of the weights in instances such as AMP. Currently you end up with a torchscript file containing 32-bit weights and script that converts specific weights to 16-bit wherever they are used.

eellison commented:
@xsacha thanks for the suggestion! Do you have a repro by any chance ?
